cpp/dolfinx/fem/Expression.h

```diff
@@ -156,34 +166,60 @@ class Expression
   /// @return The mesh
   std::shared_ptr<const mesh::Mesh> mesh() const { return _mesh; }
 
-  /// Evaluation points on the reference cell
-  /// @return Evaluation points (shape=(num_points, tdim))
-  const xt::xtensor<double, 2>& x() const { return _x; }
+  /// Get evaluation points on reference cell
+  /// @return Evaluation points
+  const xt::xtensor<double, 2>& X() const { return _X; }
 
   /// Get value size
   /// @return value_size
-  std::size_t value_size() const { return _value_size; }
+  int value_size() const
+  {
+    return std::accumulate(_value_shape.begin(), _value_shape.end(), 1,
+                           std::multiplies<int>());
+  }
+
+  /// Get value shape
+  const std::vector<int>& value_shape() const { return _value_shape; }
+
+  /// Get number of evaluation points in cell
+  /// @return number of points in cell
+  std::size_t num_points() const { return _X.shape(0); }
+
+  /// Get number of degrees-of-freedom for arguments
+  const std::vector<int>& num_argument_dofs() const
+  {
+    return _num_argument_dofs;
+  }
 
   /// Scalar type (T)
   using scalar_type = T;
 
 private:
+  // Function spaces (one for each argument)
+  std::vector<std::shared_ptr<const fem::FunctionSpace>> _function_spaces;
+
   // Coefficients associated with the Expression
   std::vector<std::shared_ptr<const fem::Function<T>>> _coefficients;
 
   // Constants associated with the Expression
   std::vector<std::shared_ptr<const fem::Constant<T>>> _constants;
 
   // Function to evaluate the Expression
-  std::function<void(T*, const T*, const T*, const double*)> _fn;
+  std::function<void(T*, const T*, const T*, const double*, const int*,
+                     const uint8_t*)>
+      _fn;
 
   // Evaluation points on reference cell
-  xt::xtensor<double, 2> _x;
+  xt::xtensor<double, 2> _X;
```

Comment:
C++ peculiarity - underscore followed by a capital is not valid C++ (https://en.cppreference.com/w/cpp/language/identifiers).

Reply:
Thanks, will change. I'll probably be learning new C++ peculiarities until the end of my days...